File: implementations/ios-sdk/shared/RichText.swift
import ContentfulOptimization
import Foundation

/// Flattens a Contentful Rich Text document into a plain display string,
/// resolving any inline merge-tag entries against the current profile.
///
/// Mirrors the React Native demo's `getRichTextContent` so the flattened text
/// (used for accessibility labels) matches byte for byte across SDKs: top-level
/// nodes are joined with a single space, a node's children with the empty string.
@MainActor
enum RichText {

    /// True when `field` is a Rich Text document node rather than a plain string.
    static func isRichTextDocument(_ field: Any?) -> Bool {
        guard let dict = field as? [String: Any] else { return false }
        return dict["nodeType"] as? String == "document" && dict["content"] is [Any]
    }

    /// Resolve an entry's `text` field to a display string: flatten a Rich Text
    /// document (resolving merge tags), pass a plain string through, otherwise
    /// fall back to `"No content"`.
    static func resolveText(_ field: Any?, client: OptimizationClient) -> String {
        if isRichTextDocument(field), let document = field as? [String: Any] {
            return flatten(document, client: client)
        }
        return field as? String ?? "No content"
    }

    /// Flatten a Rich Text document to its display string.
    static func flatten(_ document: [String: Any], client: OptimizationClient) -> String {
        guard let content = document["content"] as? [Any] else { return "" }
        return content
            .compactMap { $0 as? [String: Any] }
            .map { extractText($0, client: client) }
            .joined(separator: " ")
    }

    private static func extractText(_ node: [String: Any], client: OptimizationClient) -> String {
        switch node["nodeType"] as? String {
        case "text":
            return node["value"] as? String ?? ""
        case "embedded-entry-inline":
            return resolveEmbeddedEntry(node, client: client)
        default:
            guard let content = node["content"] as? [Any] else { return "" }
            return content
                .compactMap { $0 as? [String: Any] }
                .map { extractText($0, client: client) }
                .joined()
        }
    }

    private static func resolveEmbeddedEntry(_ node: [String: Any], client: OptimizationClient) -> String {
        guard let data = node["data"] as? [String: Any],
              let target = data["target"] as? [String: Any],
              let sys = target["sys"] as? [String: Any]
        else { return "[Merge Tag]" }

        // A still-unresolved Link means `ContentfulFetcher` did not inline the
        // entry; the flattener has nothing to resolve against.
        if sys["type"] as? String == "Link" { return "[Merge Tag]" }

        let contentTypeId = ((sys["contentType"] as? [String: Any])?["sys"] as? [String: Any])?["id"] as? String
        guard contentTypeId == "nt_mergetag" else { return "[Merge Tag]" }

        if let resolved = client.getMergeTagValue(mergeTagEntry: target), !resolved.isEmpty {
            return resolved
        }
        // Fall back to the merge tag's configured fallback value.
        let fields = target["fields"] as? [String: Any]
        return fields?["nt_fallback"] as? String ?? "[Merge Tag]"
    }
}

File: implementations/android-sdk/shared/src/main/kotlin/com/contentful/optimization/shared/RichText.kt
package com.contentful.optimization.shared

import com.contentful.optimization.core.OptimizationClient

/**
 * Flattens a Contentful Rich Text document into a plain display string,
 * resolving inline merge-tag entries against the current profile.
 *
 * Mirrors the iOS app's `RichText` so the flattened text matches byte for byte:
 * top-level nodes are joined with a single space, a node's children with the
 * empty string.
 */
@Suppress("UNCHECKED_CAST")
object RichText {

    /** True when [field] is a Rich Text document node rather than a plain string. */
    fun isRichTextDocument(field: Any?): Boolean {
        val dict = field as? Map<*, *> ?: return false
        return dict["nodeType"] == "document" && dict["content"] is List<*>
    }

    /**
     * Resolve an entry's `text` field to a display string: flatten a Rich Text
     * document (resolving merge tags), pass a plain string through, otherwise
     * fall back to `"No content"`.
     */
    suspend fun resolveText(field: Any?, client: OptimizationClient): String {
        if (isRichTextDocument(field)) {
            return flatten(field as Map<String, Any>, client)
        }
        return field as? String ?: "No content"
    }

    private suspend fun flatten(document: Map<String, Any>, client: OptimizationClient): String {
        val content = document["content"] as? List<*> ?: return ""
        val parts = mutableListOf<String>()
        for (node in content.mapNotNull { it as? Map<String, Any> }) {
            parts.add(extractText(node, client))
        }
        return parts.joinToString(" ")
    }

    private suspend fun extractText(node: Map<String, Any>, client: OptimizationClient): String {
        return when (node["nodeType"]) {
            "text" -> node["value"] as? String ?: ""
            "embedded-entry-inline" -> resolveEmbeddedEntry(node, client)
            else -> {
                val content = node["content"] as? List<*> ?: return ""
                val parts = mutableListOf<String>()
                for (child in content.mapNotNull { it as? Map<String, Any> }) {
                    parts.add(extractText(child, client))
                }
                parts.joinToString("")
            }
        }
    }

    private suspend fun resolveEmbeddedEntry(
        node: Map<String, Any>,
        client: OptimizationClient,
    ): String {
        val data = node["data"] as? Map<String, Any> ?: return "[Merge Tag]"
        val target = data["target"] as? Map<String, Any> ?: return "[Merge Tag]"
        val sys = target["sys"] as? Map<String, Any> ?: return "[Merge Tag]"

        // A still-unresolved Link means the fetcher did not inline the entry;
        // there is nothing to resolve against.
        if (sys["type"] == "Link") return "[Merge Tag]"

        val contentTypeSys =
            (sys["contentType"] as? Map<String, Any>)?.get("sys") as? Map<String, Any>
        if (contentTypeSys?.get("id") != "nt_mergetag") return "[Merge Tag]"

        val resolved = client.getMergeTagValue(target)
        if (!resolved.isNullOrEmpty()) return resolved

        // Fall back to the merge tag's configured fallback value.
        val fields = target["fields"] as? Map<String, Any>
        return fields?.get("nt_fallback") as? String ?: "[Merge Tag]"
    }
}
